### PR TITLE
Refactor the autocomplete API with Rustless

### PR DESCRIPTION
**Pull request by antoine-de**
use https://github.com/rustless/rustless/ to make the autocomplete api

endpoint
```
    ❯ http ":4000/"
HTTP/1.1 200 OK
Content-Type: application/json
Date: Thu, 21 Apr 2016 15:02:52 GMT
Transfer-Encoding: chunked

{
    "description": "autocomplete service"
}
```

v1 (need better naming, nested Response)
```
 ❯ http ":4000/v1"
HTTP/1.1 200 OK
Content-Type: application/json
Date: Thu, 21 Apr 2016 15:03:31 GMT
Transfer-Encoding: chunked

{
    "Response": {
        "description": "api version 1"
    }
}
```

wrong route (still work to do here for better return code and response):

```
 ❯ http ":4000/v1/bob"
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Thu, 21 Apr 2016 15:04:01 GMT
Transfer-Encoding: chunked

{
    "long": "bad request, error: NotMatch", 
    "short": "bad_request"
}

```

autocomplete api => error because missing param
```
 ❯ http ":4000/v1/autocomplete"
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Thu, 21 Apr 2016 15:03:53 GMT
Transfer-Encoding: chunked

{
    "long": "invalid arguments [Required { path: \"/q\" }]", 
    "short": "validation error"
}

```

real good autocomplete:
```
 ❯ http ":4000/v1/autocomplete?q=toto"
HTTP/1.1 200 OK
Content-Type: application/json
Date: Thu, 21 Apr 2016 15:05:22 GMT
Transfer-Encoding: chunked

{
    "Ok": {
        "features": [], 
        "geocoding": {
            "query": "toto", 
            "version": "0.1.0"
        }, 
        "type": "FeatureCollection"
    }
}
```
still need to do the real autocomplete stuff :stuck_out_tongue_winking_eye: 

TODO:
* [ ] real autocomplete
   * [ ] use rs-es
   * [ ] transform rs-es result into the model
* [ ] use     #[serde(skip_serializing_if="ShouldSkip::should_skip")] not to serialize None Option
* [ ] better 404
* [ ] better Result handling (do not serialize `Ok`)
* [ ] doc
* [ ] readme
* [ ] tests (use https://github.com/reem/iron-test ?)

### CODE REVIEW

File: libs/skojig/src/api.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+            endpoint.params(|params| {
+                params.req_typed("q", json_dsl::string());
+
+                params.opt("lon", |lon| {
+                    lon.coerce(json_dsl::f64());
+                    fn valid_lon(val: &json::Json,
+                                 path: &str)
+                                 -> Result<(), valico_error::ValicoErrors> {
+
+                        println!("val = {:?}, path = {}", val, path);
```
Comment (TeXitoi) [inline libs/skojig/src/api.rs]: should be removed, no?

File: libs/skojig/src/api.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                        match *val {
+                            json::Json::F64(lon) => {
+                                if lon > -180f64 && lon < 180f64 {
```
Comment (TeXitoi) [inline libs/skojig/src/api.rs]: One of the test (if not the 2) should be <=. I prefere to write that `-180. <= lon && lon <= 180.` but that's just style.

File: libs/skojig/src/api.rs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                                    return Ok(());
+                                }
+                                return Err(vec![Box::new(json_dsl::errors::WrongValue {
+                                                    path: path.to_string(),
+                                                    detail: Some("lon is not a valid coordinate"
```
Comment (TeXitoi) [inline libs/skojig/src/api.rs]: s/coordinate/longitude/

File: libs/skojig/src/api.rs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                            }
+                        }
+                    }
+
+                    lon.validate_with(valid_lon);
+                });
+                // TODO lat + check if lon then lat
+            });
+            endpoint.handle(|client, params| {
+                let q = params.find("q").unwrap().as_string().unwrap().to_string();
```
Comment (TeXitoi) [inline libs/skojig/src/api.rs]: So much unwrap :cry:

File: libs/skojig/src/lib.rs [hunk 5]
```diff
@@ -13,170 +21,42 @@
...
-}
+mod api;
+mod query;
+mod model;
 
 pub fn runserver() {
-    Iron::new(handle_query).http("0.0.0.0:3000").unwrap();
+    let api = api::root();
+    let app = Application::new(api);
+    Iron::new(app).http("localhost:4000").unwrap();
```
Comment (TeXitoi) [inline libs/skojig/src/lib.rs]: better to listen to 0.0.0.0, no?

Comment (TeXitoi) [inline libs/skojig/src/lib.rs]: or maybe just ':4000' if it works

**Comment by TeXitoi** [comment]
Seems not bad but a bit verbose (and not so functional ;))

Difficult to say without comparison.

**Comment by pbougue** [comment]
We merge it to ease coding, TODO is still todo, and we have a "little" functional regression (no autocomplete)

**Pull request merged**